### ISSUE TITLE
CTCCriterion.lua:36: assertion failed!

### ISSUE THREAD

**Issue by clarkong** [SeanNaren/CTCNetworkClient#27]
Hi @SeanNaren , I generate LMDB for AN4 dataset accroding to `prepare_an4/prepare.sh` script. Then run `th AN4CTCTrain.lua` and I get this error:

```
preparing sorted indices..
found previously saved inds..
/home/kongchang/nfs/torch/install/bin/luajit: ...ang/nfs/torch/install/share/lua/5.1/nnx/CTCCriterion.lua:36: assertion failed!
stack traceback:
        [C]: in function 'assert'
        ...ang/nfs/torch/install/share/lua/5.1/nnx/CTCCriterion.lua:36: in function 'forward'
        ./Network.lua:127: in function 'opfunc'
        .../kongchang/nfs/torch/install/share/lua/5.1/optim/sgd.lua:44: in function 'sgd'
        ./Network.lua:145: in function 'trainNetwork'
        AN4CTCTrain.lua:44: in main chunk
        [C]: in function 'dofile'
        .../nfs/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00406670
```

I locate this error in `CTCCriterion.lua:36` which is a ssert `assert(acts:nDimension() == 2)`, so I print the value which is 0. Do you have any idea?

**Comment by clarkong** [comment]
I find that:

```
local function feval(x_new)
    ......
    local predictions = self.model:forward({ inputs, sizes })
    local loss = ctcCriterion:forward(predictions, targets, sizes)
    ......
```

`predictions` 's dim is 2, and then function `ctcCriterion:forward` will call `CTCCriterion:updateOutput` which is:

```
 function CTCCriterion:updateOutput(input, target, sizes)
     assert(sizes,
         "You must pass the size of each sequence in the batch as a tensor")
     local acts = self.acts
     if input:dim() == 3 then
         acts:resizeAs(input):copy(input)
         if self.batchFirst then
             acts = acts:transpose(1, 2)
             acts = self:makeContiguous(acts)
         end 
         acts:view(acts, acts:size(1) * acts:size(2), -1)
     end 
     assert(acts:nDimension() == 2)
     self.sizes = torch.totable(sizes)
     self.gradInput = acts.new():resizeAs(acts):zero()
     if input:type() == 'torch.CudaTensor' then
         self.output = sumCosts(gpu_ctc(acts, self.gradInput, target, self.sizes))
     else
         acts = acts:float()
         self.gradInput = self.gradInput:float()
         self.output = sumCosts(cpu_ctc(acts, self.gradInput, target, self.sizes))
     end 
     return self.output
 end 
```

because of input dim is 2, so `acts:nDimension()` is 0.
now I'm trying to figure out why I get a wrong dim of `predictions`.

**Comment by clarkong** [comment]
I try to modify `function CTCCriterion:updateOutput(input, target, sizes)` like this:

```
function CTCCriterion:updateOutput(input, target, sizes)
    assert(sizes,
        "You must pass the size of each sequence in the batch as a tensor")
    local acts = self.acts
    --if input:dim() == 3 then
    if input:dim() == 2 then
        acts:resizeAs(input):copy(input)
        if self.batchFirst then
            acts = acts:transpose(1, 2)
            acts = self:makeContiguous(acts)
        end 
        --acts:view(acts, acts:size(1) * acts:size(2), -1)
        acts:view(acts, acts:size(1), acts:size(2))
    end 
    ......
end 
```

which check `input:dim() == 2` and also set `acts:view(...)`.
This works well when set nGPU = 1, but if nGPU > 1, error is returned which seems to be data parallism problem.

So, I think that maybe model definition has some problems? and that's why I get a wrong dim of `predictions`?

**Comment by SeanNaren** [comment]
Hey! So sorry for such a late reply, I'll check this out tonight and report back thanks!

**Comment by SeanNaren** [comment]
Just noticed you said this issue happens when the nGPU is set above 1. There is a bug in multi-GPU at the current stage and is being fixed in this PR  #25 !

**Comment by clarkong** [comment]
After I modify `CTCCriterion.lua` in `nnx` package, this issue goes away when nGPU == 1, then if I set nGPU > 1, new error reports.
The original problem is still existing which is wrong dimension about `predictions`(dim = 2) and it does not match `function CTCCriterion:updateOutput(input, target, sizes)` in `nnx` (input dim should be 3).
@SeanNaren

**Comment by clarkong** [comment]
fixed by reinstall torch ...
